web-api/src/models.rs

```diff
@@ -153,6 +168,68 @@ string_wrapper! {
     pub(crate) DocumentQuery, InvalidDocumentQuery, GENERIC_STRING_SYNTAX, 1..=512;
 }
 
+/// Id pointing to a specific snippet in a document.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
+pub(crate) struct SnippetId {
+    document_id: DocumentId,
+    snippet_idx: usize,
+}
+
+impl SnippetId {
+    pub(crate) fn new(document_id: DocumentId, snippet_idx: usize) -> Self {
+        Self {
+            document_id,
+            snippet_idx,
+        }
+    }
+
+    pub(crate) fn document_id(&self) -> &DocumentId {
+        &self.document_id
+    }
+
+    pub(crate) fn into_document_id(self) -> DocumentId {
+        self.document_id
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn snippet_idx(&self) -> usize {
+        self.snippet_idx
+    }
+
+    const ES_SNIPPET_ID_PREFIX: &str = "_s.";
+
+    pub(crate) fn try_from_es_id(es_id: impl AsRef<str>) -> Result<Self, Error> {
+        let es_id = es_id.as_ref();
+        if let Some(suffix) = es_id.strip_prefix(Self::ES_SNIPPET_ID_PREFIX) {
+            let Some((snippet_idx, document_id)) = suffix.split_once('.') else {
+                return Err(InternalError::from_message(format!("failed to split child id: {suffix}")).into());
+            };
+            let Ok(snippet_idx) = snippet_idx.parse() else {
+                return Err(InternalError::from_message(format!("failed to parse child idx: {snippet_idx}")).into());
+            };
+            let Ok(document_id) = document_id.parse() else {
+                return Err(InternalError::from_message(format!("failed to parse parents document id: {document_id}")).into());
+            };
+            Ok(Self::new(document_id, snippet_idx))
+        } else {
+            Ok(Self::new(es_id.parse()?, 0))
+        }
+    }
+
+    pub(crate) fn to_es_id(&self) -> Cow<'_, str> {
+        if self.snippet_idx == 0 {
+            Cow::Borrowed(&self.document_id)
```

Comment:
I really don't like this `idx == 0` special casing but we need it for backward compatibility with existing tenants.

Reply:
>  instead of snippet_idx: usize we could use a snippet_idx: Option<usize> to better transport that information. not sure if it's worth the additional option-handling effort though.

no `snippet_idx: None` as no semantic meaning at all

there is only special casing in the id used with elastic search *but not any logic*, weather a document has 1-snippet or 10-snippets it's all the same

in cases where we have a optional `snippet_idx` it's for situations where you can have either a document id or a snippet id, but that's a completely different use-case

> 0-th split part of a document

there are not splits, just documents with 1 or more snippets.

Reply:
We also only store "snippets" in elastic search. Just currently that happens to be the whole document.